cupy/core/core.pyx

```diff
@@ -1940,17 +1940,68 @@ cpdef ndarray _repeat(ndarray a, repeats, axis=None):
 
 cpdef ndarray concatenate(tup, axis, shape, dtype):
     cdef ndarray ret
+    cdef int base
     ret = ndarray(shape, dtype=dtype)
 
-    skip = (slice(None),) * axis
-    i = 0
+    all_same_type = True
     for a in tup:
-        aw = a.shape[axis]
-        ret[skip + (slice(i, i + aw),)] = a
-        i += aw
+        all_same_type &= a.dtype == tup[0].dtype
+
+    if all_same_type:
+        base = numpy.prod(shape[axis+1:], dtype='i')
+        x = array([a.data.ptr for a in tup])
```

Comment:
Please use stream.
Current code has synchronize mode copy (host to device).
Thus this code is slow.